modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/HorizontalCardTag.java

```diff
@@ -0,0 +1,383 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.taglib.clay.servlet.taglib;
+
+import com.liferay.frontend.taglib.clay.internal.servlet.taglib.BaseContainerTag;
+import com.liferay.frontend.taglib.clay.servlet.taglib.soy.HorizontalCard;
+import com.liferay.frontend.taglib.clay.servlet.taglib.util.DropdownItem;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.language.LanguageUtil;
+import com.liferay.portal.kernel.util.ListUtil;
+import com.liferay.taglib.util.TagResourceBundleUtil;
+
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.servlet.jsp.JspException;
+import javax.servlet.jsp.JspWriter;
+
+/**
+ * @author Marko Cikos
+ */
+public class HorizontalCardTag extends BaseContainerTag {
+
+	@Override
+	public int doStartTag() throws JspException {
+		setAttributeNamespace(_ATTRIBUTE_NAMESPACE);
+
+		Map<String, String> data = getData();
+
+		if (data != null) {
+			for (Map.Entry<String, String> entry : data.entrySet()) {
+				setDynamicAttribute(
+					StringPool.BLANK, "data-" + entry.getKey(),
+					entry.getValue());
+			}
+		}
+
+		return super.doStartTag();
+	}
+
+	public List<DropdownItem> getActionDropdownItems() {
+		if ((_actionDropdownItems != null) && (_horizontalCard != null)) {
+			return _horizontalCard.getActionDropdownItems();
+		}
+
+		return _actionDropdownItems;
+	}
+
+	@Override
+	public String getCssClass() {
+		if ((super.getCssClass() == null) && (_horizontalCard != null)) {
+			if (_horizontalCard.getCssClass() != null) {
+				return _horizontalCard.getCssClass();
+			}
+
+			if (_horizontalCard.getElementClasses() != null) {
+				return _horizontalCard.getElementClasses();
+			}
+		}
+
+		return super.getCssClass();
+	}
+
+	public Map<String, String> getData() {
```

Comment:
I don't think we can do this. We are doing more than dynamic attributes, we are checking for attribute setting in `_horizontalCard.getData()`.

To deprecate this safely, `data` would need to be deprecated in [`BaseClayCard` interface](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/soy/BaseClayCard.java#L39) first.

Reply:
Those are two different things:

- In the Tag we have to deprecate it. Actually, if you check `BaseContainerTag` `getData` and `setData` methods are deprecated. We don't want to support this as an attribute since we have dynamicAttributes mechanism
- In the model we don't have to deprecate it if we don't want to. We can get those data attributes from the model in the Tag `doStartTag` method and set them as dynamic attributes (as you're already doing but instead getting them from `getData()` getting them from `horizontalCard.getData()` if `horizontalCard` exists.

Reply:
Yeah, I guess it just feels weird to deprecate code that supports non-deprecated `Card` attribute.